pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go

```diff
@@ -0,0 +1,339 @@
+/*
+Copyright 2022 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/api/v1beta1"
+	batchcontext "github.com/openkruise/rollouts/pkg/controller/batchrelease/context"
+	"github.com/openkruise/rollouts/pkg/controller/batchrelease/control"
+	"github.com/openkruise/rollouts/pkg/controller/batchrelease/control/bluegreenstyle"
+	"github.com/openkruise/rollouts/pkg/controller/batchrelease/control/bluegreenstyle/hpa"
+	deploymentutil "github.com/openkruise/rollouts/pkg/controller/deployment/util"
+	"github.com/openkruise/rollouts/pkg/util"
+	"github.com/openkruise/rollouts/pkg/util/errors"
+	"github.com/openkruise/rollouts/pkg/util/patch"
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog/v2"
+	utilpointer "k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type realController struct {
+	*util.WorkloadInfo
+	client client.Client
+	pods   []*corev1.Pod
+	key    types.NamespacedName
+	object *apps.Deployment
+	finder *util.ControllerFinder
+}
+
+func NewController(cli client.Client, key types.NamespacedName, _ schema.GroupVersionKind) bluegreenstyle.Interface {
+	return &realController{
+		key:    key,
+		client: cli,
+		finder: util.NewControllerFinder(cli),
+	}
+}
+
+func (rc *realController) GetWorkloadInfo() *util.WorkloadInfo {
+	return rc.WorkloadInfo
+}
+
+func (rc *realController) BuildController() (bluegreenstyle.Interface, error) {
+	if rc.object != nil {
+		return rc, nil
+	}
+	object := &apps.Deployment{}
+	if err := rc.client.Get(context.TODO(), rc.key, object); err != nil {
+		return rc, err
+	}
+	rc.object = object
+	rc.WorkloadInfo = rc.getWorkloadInfo(object)
+	return rc, nil
+}
+
+func (rc *realController) ListOwnedPods() ([]*corev1.Pod, error) {
+	if rc.pods != nil {
+		return rc.pods, nil
+	}
+	var err error
+	rc.pods, err = util.ListOwnedPods(rc.client, rc.object)
+	return rc.pods, err
+}
+
+// Initialize prepares the Deployment for the BatchRelease process
+func (rc *realController) Initialize(release *v1beta1.BatchRelease) error {
```

Comment:
plz add ut for Initialize, UpgradeBatch and Finalize, which is the most important logic in this patch

Reply:
done